### PR TITLE
Make proxy (and UUPS) to use namespace

### LINKED ISSUE #324: Refactor UUPSImplementation to use namespace storage

### LINKED ISSUE #325: Refactor ForwardingProxy to use namespace storage

### PR DESCRIPTION
**Pull request by leomassazza**
Fix #325 
Fix #324

### CODE REVIEW

File: packages/core-contracts/contracts/proxy/ProxyStorage.sol [hunk 1]
```diff
@@ -9,8 +9,8 @@ contract ProxyStorage {
 
     function _proxyStorage() internal pure returns (ProxyNamespace storage store) {
         assembly {
-            // bytes32(uint(keccak256("io.synthetix.proxy")) - 1)
-            store.slot := 0x9dbde58b6f7305fccdc5abd7ea1096e84de3f9ee47d83d8c3efc3e5557ac9c74
+            // bytes32(uint(keccak256("io.synthetix.v3.core-contracts.proxy")) - 1)
```
Comment (leomassazza) [inline packages/core-contracts/contracts/proxy/ProxyStorage.sol:12]: np. Do you prefer just proxy or v3.core-contracts.proxy?

File: packages/core-contracts/contracts/proxy/UUPSImplementation.sol [hunk 2]
```diff
@@ -4,19 +4,28 @@ pragma solidity ^0.8.0;
...
+    function _setSimulatingUpgrade(bool simulatingUpgrade) internal virtual {
+        _proxyStorage().simulatingUpgrade = simulatingUpgrade;
+    }
 
-    function _getSimulatingUpgrade() internal view virtual returns (bool);
+    function _getSimulatingUpgrade() internal view virtual returns (bool) {
+        return _proxyStorage().simulatingUpgrade;
+    }
 
     function upgradeTo(address newImplementation) public virtual;
```
Comment (leomassazza) [inline packages/core-contracts/contracts/proxy/UUPSImplementation.sol:30]: upgradeTo is still virtual to allow the implementation to use any modifier to protect it

File: packages/core-modules/contracts/BaseProxy.sol [hunk 3]
```diff
@@ -2,18 +2,9 @@
 pragma solidity ^0.8.0;
 
 import "@synthetixio/core-contracts/contracts/proxy/ForwardingProxy.sol";
-import "./storage/ProxyStorage.sol";
 
-contract BaseProxy is ForwardingProxy, ProxyStorage {
+contract BaseProxy is ForwardingProxy {
     constructor(address firstImplementation) {
```
Comment (leomassazza) [inline packages/core-modules/contracts/BaseProxy.sol:7]: BaseProxy is the only proxy that is not going to be behind another... here the constructor works

File: packages/core-modules/contracts/modules/UpgradeModule.sol [hunk 4]
```diff
@@ -1,32 +1,15 @@
...
 import "../mixins/OwnerMixin.sol";
-import "../storage/ProxyStorage.sol";
 
-contract UpgradeModule is UniversalProxyImplementation, ProxyStorage, OwnerMixin {
+contract UpgradeModule is UUPSImplementation, OwnerMixin {
     function upgradeTo(address newImplementation) public override onlyOwner {
         _upgradeTo(newImplementation);
     }
 
     function getImplementation() public view returns (address) {
```
Comment (leomassazza) [inline packages/core-modules/contracts/modules/UpgradeModule.sol:12]: Should the public getImplementation() be in the UUPSImplementation (available for all contracts extending it or not)? I opted for not, but I'm open to suggestions

Comment (mjlescano) [reply packages/core-modules/contracts/modules/UpgradeModule.sol:12]: I like the "not" approach until necessary 👮

**Review by mjlescano — APPROVED** [review]

File: packages/core-contracts/contracts/proxy/ProxyStorage.sol (hunk 1 above)
Comment (ajsantander) [reply packages/core-contracts/contracts/proxy/ProxyStorage.sol:12]: Just proxy for now! But yes, this decision is to be finalized before going to prod.

File: packages/core-contracts/contracts/proxy/UUPSImplementation.sol (hunk 2 above)
Comment (ajsantander) [reply packages/core-contracts/contracts/proxy/UUPSImplementation.sol:30]: Yeah, I saw this earlier today. I didn't quite get the reason for having `upgradeTo` and `_upgradeTo`. As far as I understand modifiers don't affect inheritance so you don't need this. Or is there some other reason?

File: packages/core-contracts/test/proxy/ForwardingProxy.test.js [hunk 5]
```diff
@@ -13,7 +13,8 @@ describe('ForwardingProxy', () => {
       Implementation = await factory.deploy();
 
       factory = await ethers.getContractFactory('ForwardingProxyMock');
-      Proxy = await factory.deploy(Implementation.address);
+      Proxy = await factory.deploy();
+      await Proxy.initialize(Implementation.address);
```
Comment (ajsantander) [inline packages/core-contracts/test/proxy/ForwardingProxy.test.js:17]: I'd recommend to remove constructors in a different PR.

If you still want to do this, consider that this returns a tx that needs to be awaited.
```const tx = await Proxy.initialize(...)
await tx.wait()```

File: packages/core-modules/contracts/modules/UpgradeModule.sol (hunk 4 above)
Comment (ajsantander) [reply packages/core-modules/contracts/modules/UpgradeModule.sol:12]: Your call. Sounds good.

**Review by ajsantander — APPROVED** [review]
Awesome!

I'm only hesitant to include any removal of constructors in these PRs. I'd rather do them later.

**Review by i-stam — APPROVED** [review]

**Pull request merged**